x-pack/plugins/files/server/routes/file_kind/download.ts

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { schema, TypeOf } from '@kbn/config-schema';
+import { Ensure } from '@kbn/utility-types';
+import { Readable } from 'stream';
+
+import type { DownloadFileKindHttpEndpoint } from '../../../common/api_routes';
+import { fileNameWithExt } from '../common_schemas';
+import { fileErrors } from '../../file';
+import { getDownloadHeadersForFile } from '../common';
+import { getById } from './helpers';
+import type { FileKindsRequestHandler } from './types';
+
+export const method = 'get' as const;
+
+export const paramsSchema = schema.object({
+  id: schema.string(),
+  fileName: schema.maybe(fileNameWithExt),
+});
+
+type Params = Ensure<DownloadFileKindHttpEndpoint['inputs']['params'], TypeOf<typeof paramsSchema>>;
+
+type Response = Readable;
+
+export const handler: FileKindsRequestHandler<Params, unknown, Body> = async (
+  { files, fileKind },
+  req,
+  res
+) => {
+  const { fileService } = await files;
+  const {
+    params: { id, fileName },
+  } = req;
+  const { error, result: file } = await getById(fileService.asCurrentUser(), id, fileKind);
+  if (error) return error;
+  try {
+    const body: Response = await file.downloadContent();
+    return res.ok({
+      body,
+      headers: getDownloadHeadersForFile(file, fileName),
+    });
+  } catch (e) {
+    if (e instanceof fileErrors.NoDownloadAvailableError) {
+      return res.notFound({ body: { message: e.message } });
+    }
+    throw e;
```

Comment:
why not log the error and return a custom error response?

Reply:
This is a good point! This was being handled inconsistently between endpoints.

I think the simplest way to handle this would be to always `throw` because that will result in a `500` being returned with an opaque "internal server error" message.

The benefit of doing it this way is that we don't send internal server messages over the wire and rather treat those cases as really unexpected errors. By throwing them, we also know they will appear in server logs as errors.

The only exception to this is handling "domain-specific" errors in the controllers. The default case with this should still be to just re-throw the error though.

Let me know what you think!

Reply:
This sounds like what I was pushing for! Thanks JL